RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegAlgoRR.cc

```diff
@@ -0,0 +1,319 @@
+      float Dtheta_min_l2 = 999;
+     
+      std::vector<const GEMRecHit*>::const_iterator grhIt = gemrechits.begin();
+      const GEMRecHit* gemrh_min_l1= *grhIt;
+      const GEMRecHit* gemrh_min_l2= *grhIt;
+      
+      // Loop over GEM rechits from the EnsembleGEMHitContainer
+      for(grhIt = gemrechits.begin(); grhIt != gemrechits.end(); ++grhIt) 
+	{
+	
+	  // get GEM Rechit Local & Global Position
+	  auto rhLP = (*grhIt)->localPosition();
+	  const GEMEtaPartition* rhRef  = theGEMEtaParts_.find((*grhIt)->gemId())->second;
+	  auto rhGP = rhRef->toGlobal(rhLP);
+	  // get GEM Rechit Local position w.r.t. the CSC Chamber
+	  // --> we are interessed in the z-coordinate
+	  auto rhLP_inSegmRef = cscChamber->toLocal(rhGP);
+	  // calculate the extrapolation of the CSC segment to the GEM plane (z-coord)
+	  // to get x- and y- coordinate 
+	  float xe = segLP.x()+segLD.x()*rhLP_inSegmRef.z()/segLD.z();
+	  float ye = segLP.y()+segLD.y()*rhLP_inSegmRef.z()/segLD.z();
+	  float ze = rhLP_inSegmRef.z();
+	  // 3D extrapolated point in the GEM plane
+	  LocalPoint extrPoint(xe,ye,ze);
+	  
+	  // get GEM Rechit Global Position, but obtained from CSC Chamber
+	  // --> this should be the same as rhGP = rhRef->toGlobal(rhLP);
+	  auto rhGP_fromSegmRef = cscChamber->toGlobal(rhLP_inSegmRef);
+	  float phi_rh          = rhGP_fromSegmRef.phi();
+	  float theta_rh        = rhGP_fromSegmRef.theta();
+	  // get Extrapolat Global Position, also obtained from CSC Chamber
+	  auto extrPoinGP_fromSegmRef = cscChamber->toGlobal(extrPoint);
+	  float phi_ext               = extrPoinGP_fromSegmRef.phi();
+	  float theta_ext             = extrPoinGP_fromSegmRef.theta();
+	  
+	  // GEM 1st Layer :: Search for GEM Rechit with smallest Delta Eta and Delta Phi
+	  //                  and save it inside gemrh_min_l1 
+	  if ((*grhIt)->gemId().layer()==1) 
+	    {
+	      float Dphi_l1 = fabs(phi_ext-phi_rh);
+	      float Dtheta_l1 = fabs(theta_ext-theta_rh);
+	      if (Dphi_l1 <= Dphi_min_l1) 
+		{
+		  Dphi_min_l1   = Dphi_l1;
+		  Dtheta_min_l1 = Dtheta_l1;
+		  gemrh_min_l1  = *grhIt;
+		}	  
+	    }
+	  // GEM 2nd Layer :: Search for GEM Rechit with smallest Delta Eta and Delta Phi
+	  //                  and save it inside gemrh_min_l2 
+	  if ((*grhIt)->gemId().layer()==2) 
+	    {
+	      float Dphi_l2 = fabs(phi_ext-phi_rh);
+	      float Dtheta_l2 = fabs(theta_ext-theta_rh);
+	      if (Dphi_l2 <= Dphi_min_l2)
+		{
+		  Dphi_min_l2   = Dphi_l2;
+		  Dtheta_min_l2 = Dtheta_l2;
+		  gemrh_min_l2  = *grhIt;
+		}
+	    }
+
+	} // end loop over GEM Rechits
+      
+      // Check whether GEM rechit with smallest delta eta and delta phi 
+      // w.r.t. the extrapolation of the CSC segment is within the 
+      // maxima given by the configuration of the algorithm
+      bool phiRequirementOK_l1 = Dphi_min_l1 < dPhiChainBoxMax;
+      bool thetaRequirementOK_l1 = Dtheta_min_l1 < dThetaChainBoxMax;
+      if(phiRequirementOK_l1 && thetaRequirementOK_l1) 
```

Comment:
technically that is not possible, because of the comparison between a pointer and an iterator is not allowed. I added an additional check that the pointer is not NULL: 
gemrh_min_l1!=0
but in principle the protection is already implemented in line 184:
if(gemrechits.size()!=0)
this avoids that when initializing gemrh_min_l1 at line 192:
const GEMRecHit* gemrh_min_l1= *grhIt;
can result in the fact that it actually gets this value: gemrh_min_l1= *(gemrechits.end())
because if gemrechits.begin() == gemrechits.end(), then gemrechits.size() = 0;